framework/Databases.py

```diff
@@ -222,402 +243,87 @@ def getEndingGroupNames(self):
     endingGroups = self.database.retrieveAllHistoryNames()
     return endingGroups
 
-  def addGroup(self,attributes,loadFrom,upGroup=False):
+  def addRealization(self,rlz):
     """
-      Function used to add a group to the database
-      @ In, attributes, dict, options
-      @ In, loadFrom, string, source of the data
-      @ In, upGroup, bool, optional, the group (if present) needs to be updated?
+      Adds a "row" (or "sample") to this database.
+      This is the method to add data to this database.
+      Note that rlz can include many more variables than this database actually wants.
+      Before actually adding the realization, data is formatted for this data object.
+      @ In, rlz, dict, {var:val} format where
+                         "var" is the variable name as a string,
+                         "val" is either a float or a np.ndarray of values.
       @ Out, None
     """
-    if 'metadata' in attributes.keys():
-      attributes['group'] = attributes['metadata']['prefix']
-    elif 'prefix' in attributes.keys():
-      attributes['group'] = attributes['prefix']
-    else:
-      self.raiseAnError(IOError,'addGroup function needs a prefix (ID) for adding a new group to a database!')
-    self.database.addGroup(attributes['group'],attributes,loadFrom,upGroup)
+    # realization must be a dictionary
+    assert(type(rlz).__name__ == "dict")
+    # prefix must be present
+    assert('prefix' in rlz)
+    self.database.addGroup(rlz)
     self.built = True
 
-  def addGroupDataObjects(self,attributes,loadFrom,upGroup=False):
+  def addExpectedMeta(self,keys):
     """
-      Function to add a group in the HDF5 database
-      @ In, attributes, dict, options
-      @ In, loadFrom, DataObjects or dict, source of the data
-      @ In, upGroup, bool, optional, the group (if present) needs to be updated?
+      Registers meta to look for in realizations.
+      @ In, keys, set(str), keys to register
       @ Out, None
     """
-    source = {}
-    if type(loadFrom) != dict:
-      if not loadFrom.type in ['PointSet','HistorySet']:
-        self.raiseAnError(IOError,'addGroupDataObjects function needs to have a Data(s) as input source')
-      source['type'] = 'DataObjects'
-    source['name'] = loadFrom
-    self.database.addGroupDataObjects(attributes['group'],attributes,source,upGroup,specificVars=self.variables)
-    self.built = True
+    self.database.addExpectedMeta(keys)
+    self.addMetaKeys(*keys)
 
-  def initialize(self,gname,attributes=None,upGroup=False):
+  def initialize(self,gname,options={}):
```

Comment:
It's not safe to set defaults to empty dictionaries; that dictionary can be changed and the default is changed for every subsequent call (see https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html).

Safer is to set it to `None` and if it's none, then set it to a dict.

Reply:
Done